File: Sources/ArtifactsView.swift
import SwiftUI

/// Everything the agents produce — organized any way you think: by project,
/// repo, day, or type; filtered by kind, size, and lifespan; sorted in one
/// tap. Purpose decides lifespan: ephemeral artifacts carry a TTL and file
/// themselves into the archive; permanent ones keep until superseded.
struct ArtifactsView: View {
    @EnvironmentObject var store: AppStore
    @Environment(\.colorScheme) private var scheme

    @State private var kindFilter: ArtifactKind? = nil
    @State private var grouping = "Project"
    @State private var sizeBand = "Any size"
    @State private var lifeFilter = "Any life"
    @State private var sort = "Newest"

    // MARK: Pipeline — one pass over the gallery per body, not one per control

    private struct GalleryModel {
        var sections: [(header: String, items: [Artifact])] = []
        var kindCounts: [ArtifactKind: Int] = [:]
        var total = 0
        var visible = 0
    }

    private var model: GalleryModel {
        var out = GalleryModel()
        out.total = store.artifacts.count
        var items: [Artifact] = []
        items.reserveCapacity(out.total)
        for a in store.artifacts {
            out.kindCounts[a.kind, default: 0] += 1
            if let kindFilter, a.kind != kindFilter { continue }
            switch sizeBand {
            case "< 100 KB": if a.sizeKB >= 100 { continue }
            case "100 KB – 1 MB": if a.sizeKB < 100 || a.sizeKB >= 1024 { continue }
            case "> 1 MB": if a.sizeKB < 1024 { continue }
            default: break
            }
            switch lifeFilter {
            case "Permanent": if !a.life.isPermanent { continue }
            case "Ephemeral": if a.life.isPermanent { continue }
            default: break
            }
            items.append(a)
        }
        switch sort {
        case "Largest": items.sort { $0.sizeKB > $1.sizeKB }
        case "A–Z": items.sort { $0.title < $1.title }
        default: break // Newest = feed order
        }
        out.visible = items.count
        let key: ((Artifact) -> String)?
        switch grouping {
        case "Repo": key = { $0.repo }
        case "Day": key = { $0.day }
        case "Type": key = { $0.kind.rawValue + "s" }
        case "Project": key = { $0.room }
        default: key = nil
        }
        guard let key else {
            out.sections = [("", items)]
            return out
        }
        var order: [String] = []
        var buckets: [String: [Artifact]] = [:]
        for item in items {
            let k = key(item)
            if buckets[k] == nil { order.append(k) }
            buckets[k, default: []].append(item)
        }
        out.sections = order.map { ($0, buckets[$0] ?? []) }
        return out
    }

    var body: some View {
        let model = model
        ScrollView {
            VStack(alignment: .leading, spacing: 0) {
                kindChips(model).padding(.top, 8)
                controlBar.padding(.top, 10)

                ForEach(model.sections, id: \.header) { section in
                    if !section.header.isEmpty {
                        HStack {
                            Text(section.header)
                                .font(.system(size: 12, weight: .heavy))
                                .foregroundStyle(DT.ink55(scheme))
                            Spacer()
                            Text("\(section.items.count)")
                                .font(.system(size: 10, weight: .bold, design: .monospaced))
                                .foregroundStyle(DT.ink35(scheme))
                        }
                        .padding(.top, 18)
                    }
                    LazyVGrid(columns: [GridItem(.flexible(), spacing: 9), GridItem(.flexible())], spacing: 9) {
                        ForEach(section.items) { artifact in
                            NavigationLink { ArtifactDetailView(artifactId: artifact.id) } label: {
                                ArtifactCard(artifact: artifact)
                            }
                            .buttonStyle(Pressable())
                        }
                    }
                    .padding(.top, 9)
                }

                if model.visible == 0 {
                    VStack(spacing: 8) {
                        Image(systemName: "tray")
                            .font(.system(size: 26, weight: .light))
                            .foregroundStyle(DT.ink35(scheme))
                        Text("Nothing matches those filters")
                            .font(.system(size: 13, weight: .semibold))
                            .foregroundStyle(DT.ink55(scheme))
                    }
                    .frame(maxWidth: .infinity)
                    .padding(.vertical, 60)
                }

                Text("Ephemeral artifacts file to the archive when their TTL passes — searchable, never deleted. Permanent ones keep until superseded.")
                    .font(.system(size: 10.5))
                    .foregroundStyle(DT.ink55(scheme))
                    .multilineTextAlignment(.center)
                    .frame(maxWidth: .infinity)
                    .padding(.top, 22)
                Spacer(minLength: 24)
            }
            .padding(.horizontal, 20)
        }
        .background(DT.page(scheme).ignoresSafeArea())
        .navigationTitle("Artifacts")
        .navigationBarTitleDisplayMode(.inline)
        .onAppear { store.intent.record(.artifacts) }
        .animation(.spring(response: 0.35, dampingFraction: 0.85), value: grouping)
        .animation(.spring(response: 0.35, dampingFraction: 0.85), value: sort)
    }

    // MARK: Controls

    private func kindChips(_ model: GalleryModel) -> some View {
        ScrollView(.horizontal, showsIndicators: false) {
            HStack(spacing: 7) {
                chip(nil, label: "All · \(model.total)")
                ForEach(ArtifactKind.allCases, id: \.self) { kind in
                    if let count = model.kindCounts[kind], count > 0 {
                        chip(kind, label: "\(kind.rawValue) · \(count)")
                    }
                }
            }
        }
        .padding(.horizontal, -20)
        .contentMargins(.horizontal, 20, for: .scrollContent)
    }

    private func chip(_ kind: ArtifactKind?, label: String) -> some View {
        let selected = kindFilter == kind
        let tint = kind?.tint ?? DT.violet
        return Button {
            withAnimation(.spring(response: 0.35, dampingFraction: 0.85)) { kindFilter = kind }
        } label: {
            HStack(spacing: 6) {
                if let kind {
                    Image(systemName: kind.symbol).font(.system(size: 10, weight: .semibold))
                }
                Text(label).font(.system(size: 12.5, weight: .bold))
            }
            .foregroundStyle(selected ? .white : DT.ink78(scheme))
            .padding(.horizontal, 12).padding(.vertical, 8)
            .background(selected ? AnyShapeStyle(tint) : AnyShapeStyle(DT.surface(scheme)))
            .clipShape(Capsule())
            .overlay(Capsule().strokeBorder(selected ? .clear : DT.hairline(scheme), lineWidth: 0.8))
        }
        .buttonStyle(Pressable())
    }

    private var controlBar: some View {
        ScrollView(.horizontal, showsIndicators: false) {
            HStack(spacing: 7) {
                menuPill(icon: "square.grid.2x2", selection: $grouping,
                         options: ["Project", "Repo", "Day", "Type", "None"], prefix: "By")
                menuPill(icon: "internaldrive", selection: $sizeBand,
                         options: ["Any size", "< 100 KB", "100 KB – 1 MB", "> 1 MB"], prefix: nil)
                menuPill(icon: "hourglass", selection: $lifeFilter,
                         options: ["Any life", "Permanent", "Ephemeral"], prefix: nil)
                menuPill(icon: "arrow.up.arrow.down", selection: $sort,
                         options: ["Newest", "Largest", "A–Z"], prefix: nil)
            }
        }
        .padding(.horizontal, -20)
        .contentMargins(.horizontal, 20, for: .scrollContent)
    }

    private func menuPill(icon: String, selection: Binding<String>, options: [String], prefix: String?) -> some View {
        Menu {
            Picker("", selection: selection) {
                ForEach(options, id: \.self) { Text($0).tag($0) }
            }
        } label: {
            HStack(spacing: 5) {
                Image(systemName: icon).font(.system(size: 10, weight: .semibold))
                Text(prefix.map { "\($0) \(selection.wrappedValue.lowercased())" } ?? selection.wrappedValue)
                    .font(.system(size: 11.5, weight: .bold))
                Image(systemName: "chevron.down").font(.system(size: 8, weight: .bold))
            }
            .foregroundStyle(selection.wrappedValue.hasPrefix("Any") || selection.wrappedValue == "Newest" || selection.wrappedValue == "None"
                             ? DT.ink55(scheme) : DT.violetText(scheme))
            .padding(.horizontal, 11).padding(.vertical, 8)
            .background(DT.surface2(scheme))
            .clipShape(Capsule())
        }
    }
}

struct ArtifactCard: View {
    @EnvironmentObject var store: AppStore
    @Environment(\.colorScheme) private var scheme
    let artifact: Artifact

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            HStack {
                Image(systemName: artifact.kind.symbol)
                    .font(.system(size: 14, weight: .semibold))
                    .foregroundStyle(artifact.kind.tint)
                    .frame(width: 30, height: 30)
                    .background(artifact.kind.tint.opacity(0.14))
                    .clipShape(RoundedRectangle(cornerRadius: 9, style: .continuous))
                Spacer()
                Text(artifact.age)
                    .font(.system(size: 10, design: .monospaced))
                    .foregroundStyle(DT.ink35(scheme))
            }
            Text(artifact.title)
                .font(.system(size: 13.5, weight: .bold))
                .foregroundStyle(DT.ink(scheme))
                .lineLimit(2, reservesSpace: true)
                .multilineTextAlignment(.leading)
                .padding(.top, 10)
            Text(artifact.meta)
                .font(.system(size: 10.5, design: .monospaced))
                .foregroundStyle(artifact.kind.tint)
                .lineLimit(1)
                .padding(.top, 3)
            HStack(spacing: 6) {
                AvatarChip(letter: String(artifact.agentName.prefix(1)), color: artifact.agentColor, size: 16)
                Text(artifact.room)
                    .font(.system(size: 10))
                    .foregroundStyle(DT.ink55(scheme))
                    .lineLimit(1)
            }
            .padding(.top, 9)
            HStack {
                Text(artifact.sizeLabel)
                    .font(.system(size: 9.5, weight: .semibold, design: .monospaced))
                    .foregroundStyle(DT.ink35(scheme))
                Spacer()
                HStack(spacing: 4) {
                    Image(systemName: artifact.life.symbol)
                        .font(.system(size: 8, weight: .bold))
                    Text(artifact.life.badge)
                        .font(.system(size: 9, weight: .heavy))
                }
                .foregroundStyle(artifact.life.isPermanent ? DT.ink55(scheme) : Color(hex: 0xFFC55C))
                .padding(.horizontal, 7).padding(.vertical, 3.5)
                .background(artifact.life.isPermanent ? AnyShapeStyle(DT.surface2(scheme))
                                                      : AnyShapeStyle(Color(hex: 0xFFC55C).opacity(0.13)))
                .clipShape(Capsule())
            }
            .padding(.top, 9)
        }
        .padding(12)
        .background(DT.surface(scheme))
        .clipShape(RoundedRectangle(cornerRadius: DT.rCard, style: .continuous))
        .overlay(RoundedRectangle(cornerRadius: DT.rCard, style: .continuous)
            .strokeBorder(DT.hairline(scheme), lineWidth: 0.8))
        .contextMenu {
            Button { store.joinCall() } label: { Label("Open in session", systemImage: "waveform") }
            ShareLink(item: "\(artifact.title) — \(artifact.meta) · \(artifact.room)") {
                Label("Share", systemImage: "square.and.arrow.up")
            }
            Divider()
            if artifact.life.isPermanent {
                Button { store.setArtifactLife(artifact.id, .ephemeral(daysLeft: 7)) } label: {
                    Label("Make ephemeral · 7d", systemImage: "hourglass")
                }
                Button { store.setArtifactLife(artifact.id, .ephemeral(daysLeft: 30)) } label: {
                    Label("Make ephemeral · 30d", systemImage: "hourglass.bottomhalf.filled")
                }
            } else {
                Button { store.setArtifactLife(artifact.id, .permanent) } label: {
                    Label("Mark permanent", systemImage: "infinity")
                }
            }
        }
        .accessibilityElement(children: .ignore)
        .accessibilityLabel("\(artifact.kind.rawValue): \(artifact.title), \(artifact.room), \(artifact.sizeLabel), \(artifact.life.isPermanent ? "permanent" : "ephemeral, \(artifact.life.badge)")")
        .accessibilityHint("Opens the artifact. Long press for lifecycle options.")
    }
}

/// The Home rail — the freshest work products, one swipe deep.
struct ArtifactRail: View {
    @EnvironmentObject var store: AppStore
    @Environment(\.colorScheme) private var scheme

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            HStack {
                Eyebrow("ARTIFACTS")
                Spacer()
                NavigationLink { ArtifactsView() } label: {
                    HStack(spacing: 4) {
                        Text("All \(store.artifacts.count)")
                            .font(.system(size: 12, weight: .bold))
                        Image(systemName: "chevron.right")
                            .font(.system(size: 9, weight: .bold))
                    }
                    .foregroundStyle(DT.violetText(scheme))
                }
                .buttonStyle(Pressable())
            }
            ScrollView(.horizontal, showsIndicators: false) {
                LazyHStack(spacing: 9) {
                    ForEach(store.artifacts.prefix(6)) { artifact in
                        NavigationLink { ArtifactDetailView(artifactId: artifact.id) } label: {
                            railCard(artifact)
                        }
                        .buttonStyle(Pressable())
                    }
                }
            }
            .frame(height: 96)
            .padding(.top, 10)
            .padding(.horizontal, -20)
            .contentMargins(.horizontal, 20, for: .scrollContent)
        }
    }

    private func railCard(_ artifact: Artifact) -> some View {
        VStack(alignment: .leading, spacing: 7) {
            HStack(spacing: 6) {
                Image(systemName: artifact.kind.symbol)
                    .font(.system(size: 11, weight: .semibold))
                    .foregroundStyle(artifact.kind.tint)
                Text(artifact.kind.rawValue.uppercased())
                    .font(.system(size: 8.5, weight: .heavy))
                    .tracking(0.8)
                    .foregroundStyle(artifact.kind.tint)
                Spacer()
                if !artifact.life.isPermanent {
                    Image(systemName: "hourglass")
                        .font(.system(size: 8, weight: .bold))
                        .foregroundStyle(Color(hex: 0xFFC55C))
                }
            }
            Text(artifact.title)
                .font(.system(size: 12.5, weight: .bold))
                .foregroundStyle(DT.ink(scheme))
                .lineLimit(2)
                .multilineTextAlignment(.leading)
            Text("\(artifact.room) · \(artifact.age)")
                .font(.system(size: 9.5))
                .foregroundStyle(DT.ink55(scheme))
                .lineLimit(1)
        }
        .padding(11)
        .frame(width: 168, alignment: .leading)
        .background(DT.surface(scheme))
        .clipShape(RoundedRectangle(cornerRadius: DT.rCard, style: .continuous))
        .overlay(RoundedRectangle(cornerRadius: DT.rCard, style: .continuous)
            .strokeBorder(artifact.kind.tint.opacity(0.22), lineWidth: 0.8))
    }
}

File: web/src/views/tasks/ArtifactsView.js
// Everything the agents produce — a port of `Sources/ArtifactsView.swift`.
// Organized any way you think: by project, repo, day, or type; filtered by
// kind, size, and lifespan; sorted in one tap. Purpose decides lifespan:
// ephemeral artifacts carry a TTL and file themselves into the archive;
// permanent ones keep until superseded. One pass over the gallery per render.
import { html, cx, useState, useMemo, useObservable, useLongPress, haptic } from "../../ui.js";
import { nav } from "../../nav.js";
import { Icon, AvatarChip, Screen, Empty } from "../../components.js";
import { ARTIFACT_KINDS, ARTIFACT_META, lifeBadge, lifeSymbol, sizeLabel, permanent, ephemeral } from "../../models.js";
import { shareText } from "./shared.js";

let store = null;
export function bindArtifactsStore(s) { store = s; }

const GROUPINGS = ["Project", "Repo", "Day", "Type", "None"];
const SIZE_BANDS = ["Any size", "< 100 KB", "100 KB – 1 MB", "> 1 MB"];
const LIFE_FILTERS = ["Any life", "Permanent", "Ephemeral"];
const SORTS = ["Newest", "Largest", "A–Z"];

export const shareLine = (a) => `${a.title} — ${a.meta} · ${a.room}`;

/** Context-menu items shared by the gallery card and the Home rail. */
export function artifactMenuItems(a, { lifecycle = true } = {}) {
  const items = [
    { label: "Open in session", icon: "waveform", onSelect: () => store.joinCall() },
    { label: "Share", icon: "square.and.arrow.up", onSelect: () => shareText(shareLine(a)) },
  ];
  if (!lifecycle) return items;
  if (a.life.permanent) {
    items.push({ label: "Make ephemeral · 7d", icon: "hourglass", onSelect: () => store.setArtifactLife(a.id, ephemeral(7)) });
    items.push({ label: "Make ephemeral · 30d", icon: "hourglass.bottomhalf.filled", onSelect: () => store.setArtifactLife(a.id, ephemeral(30)) });
  } else items.push({ label: "Mark permanent", icon: "infinity", onSelect: () => store.setArtifactLife(a.id, permanent()) });
  return items;
}

// ---------------------------------------------------------- pipeline

/** One pass: counts per kind, filters, sort, grouping. */
function buildGallery(artifacts, { kindFilter, grouping, sizeBand, lifeFilter, sort }) {
  const out = { sections: [], kindCounts: {}, total: artifacts.length, visible: 0 };
  const items = [];
  for (const a of artifacts) {
    out.kindCounts[a.kind] = (out.kindCounts[a.kind] ?? 0) + 1;
    if (kindFilter && a.kind !== kindFilter) continue;
    if (sizeBand === "< 100 KB" && a.sizeKB >= 100) continue;
    if (sizeBand === "100 KB – 1 MB" && (a.sizeKB < 100 || a.sizeKB >= 1024)) continue;
    if (sizeBand === "> 1 MB" && a.sizeKB < 1024) continue;
    if (lifeFilter === "Permanent" && !a.life.permanent) continue;
    if (lifeFilter === "Ephemeral" && a.life.permanent) continue;
    items.push(a);
  }
  if (sort === "Largest") items.sort((a, b) => b.sizeKB - a.sizeKB);
  else if (sort === "A–Z") items.sort((a, b) => a.title.localeCompare(b.title));
  out.visible = items.length;
  const key = grouping === "Repo" ? (a) => a.repo : grouping === "Day" ? (a) => a.day : grouping === "Type" ? (a) => `${a.kind}s` : grouping === "Project" ? (a) => a.room : null;
  if (!key) { out.sections = [{ header: "", items }]; return out; }
  const order = [], buckets = new Map();
  for (const a of items) { const k = key(a); if (!buckets.has(k)) { order.push(k); buckets.set(k, []); } buckets.get(k).push(a); }
  out.sections = order.map((h) => ({ header: h, items: buckets.get(h) }));
  return out;
}

// -------------------------------------------------------------- view

export function ArtifactsView() {
  const s = useObservable(store);
  const [kindFilter, setKindFilter] = useState(null);
  const [grouping, setGrouping] = useState("Project");
  const [sizeBand, setSizeBand] = useState("Any size");
  const [lifeFilter, setLifeFilter] = useState("Any life");
  const [sort, setSort] = useState("Newest");
  // Lifecycle edits mutate artifacts in place; a cheap signature keeps the memo honest.
  const lifeSig = s.artifacts.map((a) => (a.life.permanent ? "p" : a.life.daysLeft)).join(",");
  const model = useMemo(() => buildGallery(s.artifacts, { kindFilter, grouping, sizeBand, lifeFilter, sort }), [s.artifacts, s.artifacts.length, lifeSig, kindFilter, grouping, sizeBand, lifeFilter, sort]);

  if (!s.artifacts.length) {
    return html`<${Screen} title="Artifacts" back contentClass="tk-pushed">
      <${Empty} icon="tray" title="No artifacts yet" body=${s.configuration.previewContentEnabled ? "Plans, diffs, reports and replays land here as agents publish them." : "Artifacts arrive from the writer once a room is live. Nothing is seeded in this build."} />
    </${Screen}>`;
  }

  return html`<${Screen} title="Artifacts" back contentClass="tk-pushed">
    <div class="tk-hscroll tk-chips" role="tablist" aria-label="Filter by kind">
      <${KindChip} kind=${null} label=${`All · ${model.total}`} selected=${kindFilter === null} onSelect=${() => setKindFilter(null)} />
      ${ARTIFACT_KINDS.map((k) => (model.kindCounts[k] ? html`<${KindChip} key=${k} kind=${k} label=${`${k} · ${model.kindCounts[k]}`} selected=${kindFilter === k} onSelect=${() => setKindFilter(k)} />` : null))}
    </div>
    <div class="tk-hscroll tk-ctl" role="group" aria-label="Group, filter and sort">
      <${MenuPill} icon="square.grid.2x2" value=${grouping} options=${GROUPINGS} prefix="By" onChange=${setGrouping} title="Group by" />
      <${MenuPill} icon="server.rack" value=${sizeBand} options=${SIZE_BANDS} onChange=${setSizeBand} title="Size" />
      <${MenuPill} icon="hourglass" value=${lifeFilter} options=${LIFE_FILTERS} onChange=${setLifeFilter} title="Lifespan" />
      <${MenuPill} icon="arrow.up.arrow.down" value=${sort} options=${SORTS} onChange=${setSort} title="Sort" />
    </div>
    ${model.sections.map((sec) => html`<div key=${sec.header || "all"}>
      ${sec.header ? html`<div class="tk-sechead"><span>${sec.header}</span><span class="tk-n">${sec.items.length}</span></div>` : null}
      <div class="tk-agrid" role="list">${sec.items.map((a) => html`<${ArtifactCard} key=${a.id} artifact=${a} />`)}</div>
    </div>`)}
    ${model.visible === 0 ? html`<div class="tk-nomatch"><${Icon} name="tray" size=${26} weight=${1.4} color="var(--ink-35)" />Nothing matches those filters</div>` : null}
    <p class="tk-note">Ephemeral artifacts file to the archive when their TTL passes — searchable, never deleted. Permanent ones keep until superseded.</p>
  </${Screen}>`;
}

function KindChip({ kind, label, selected, onSelect }) {
  const tint = kind ? ARTIFACT_META[kind].tint : "var(--violet)";
  return html`<button type="button" role="tab" aria-selected=${selected} class=${cx("tk-chip pressable", selected && "tk-on")} style=${{ "--tint": tint }} onClick=${() => { if (!selected) { haptic("light"); onSelect(); } }}>
    ${kind ? html`<${Icon} name=${ARTIFACT_META[kind].symbol} size=${12} weight=${2.4} />` : null}${label}
  </button>`;
}

/** Menu pill: the Swift `Menu { Picker }` — opens a checked menu at the pill. */
function MenuPill({ icon, value, options, prefix, onChange, title }) {
  const neutral = value.startsWith("Any") || value === "Newest" || value === "None";
  const open = (e) => {
    const r = e.currentTarget.getBoundingClientRect();
    nav.openMenu({ x: r.left, y: r.bottom, title, items: options.map((o) => ({ label: o, checked: o === value, onSelect: () => onChange(o) })) });
  };
  return html`<button type="button" class=${cx("tk-mpill pressable", !neutral && "tk-set")} aria-haspopup="menu" aria-label=${`${title}: ${value}`} onClick=${(e) => { haptic("light"); open(e); }}>
    <${Icon} name=${icon} size=${11} weight=${2.4} />${prefix ? `${prefix} ${value.toLowerCase()}` : value}<${Icon} name="chevron.down" size=${9} weight=${3} />
  </button>`;
}

export function ArtifactCard({ artifact: a }) {
  const meta = ARTIFACT_META[a.kind];
  const eph = !a.life.permanent;
  const lp = useLongPress((e, pt) => nav.openMenu({ x: pt.x, y: pt.y, title: a.title, items: artifactMenuItems(a) }), { onClick: () => { haptic("light"); nav.push("artifact", { id: a.id }); } });
  return html`<button type="button" role="listitem" class="tk-art pressable" style=${{ "--tint": meta.tint }} ...${lp}
    aria-label=${`${a.kind}: ${a.title}, ${a.room}, ${sizeLabel(a.sizeKB)}, ${a.life.permanent ? "permanent" : `ephemeral, ${lifeBadge(a.life)}`}`} title="Opens the artifact. Long press for lifecycle options.">
    <div class="tk-top"><span class="tk-kind"><${Icon} name=${meta.symbol} size=${15} weight=${2.3} /></span><span class="tk-age">${a.age}</span></div>
    <div class="tk-title">${a.title}</div>
    <div class="tk-art-meta">${a.meta}</div>
    <div class="tk-who"><${AvatarChip} letter=${a.agentName.charAt(0)} color=${a.agentColor} size=${16} /><span>${a.room}</span></div>
    <div class="tk-art-foot">
      <span class="tk-size">${sizeLabel(a.sizeKB)}</span>
      <span class=${cx("tk-life", eph && "tk-eph")}><${Icon} name=${lifeSymbol(a.life)} size=${9} weight=${2.8} />${lifeBadge(a.life)}</span>
    </div>
  </button>`;
}

/** The Home rail — the freshest work products, one swipe deep (exported for the Home port). */
export function ArtifactRail() {
  const s = useObservable(store);
  if (!s.artifacts.length) return null;
  return html`<div>
    <div class="hstack" style=${{ justifyContent: "space-between" }}>
      <div class="eyebrow">ARTIFACTS</div>
      <button type="button" class="pressable hstack t-sm w7" style=${{ color: "var(--violet-text)", gap: 4, minHeight: 32 }} onClick=${() => { haptic("light"); nav.push("artifacts"); }}>All ${s.artifacts.length}<${Icon} name="chevron.right" size=${10} weight=${3} /></button>
    </div>
    <div class="tk-hscroll" style=${{ marginTop: 10, gap: 9 }} role="list">
      ${s.artifacts.slice(0, 6).map((a) => html`<${RailCard} key=${a.id} artifact=${a} />`)}
    </div>
  </div>`;
}

function RailCard({ artifact: a }) {
  const meta = ARTIFACT_META[a.kind];
  const lp = useLongPress((e, pt) => nav.openMenu({ x: pt.x, y: pt.y, preview: html`<${RailPeek} artifact=${a} />`, items: artifactMenuItems(a, { lifecycle: false }) }), { onClick: () => { haptic("light"); nav.push("artifact", { id: a.id }); } });
  return html`<button type="button" role="listitem" class="tk-railcard pressable" style=${{ "--tint": meta.tint }} ...${lp} aria-label=${`${a.kind}: ${a.title}, ${a.room}, ${a.age} ago`}>
    <div class="hstack" style=${{ gap: 6 }}>
      <${Icon} name=${meta.symbol} size=${12} weight=${2.4} color=${meta.tint} /><span class="eyebrow" style=${{ fontSize: 8.5, letterSpacing: .8, color: meta.tint }}>${a.kind}</span>
      <span class="grow" />${!a.life.permanent ? html`<${Icon} name="hourglass" size=${9} weight=${2.8} color="var(--tint-amber)" />` : null}
    </div>
    <div class="clamp2 w7" style=${{ fontSize: 12.5, lineHeight: 1.25 }}>${a.title}</div>
    <div class="truncate" style=${{ fontSize: 9.5, color: "var(--ink-55)" }}>${a.room} · ${a.age}</div>
  </button>`;
}

function RailPeek({ artifact: a }) {
  const meta = ARTIFACT_META[a.kind];
  return html`<div class="tk-peek" style=${{ "--tint": meta.tint }}>
    <div class="hstack" style=${{ gap: 10 }}>
      <span class="tk-kind" style=${{ width: 36, height: 36, borderRadius: 10 }}><${Icon} name=${meta.symbol} size=${16} weight=${2.3} /></span>
      <div class="grow" style=${{ minWidth: 0 }}><div class="w8" style=${{ fontSize: 15 }}>${a.title}</div><div class="mono" style=${{ fontSize: 11, color: meta.tint }}>${a.meta}</div></div>
    </div>
    <div class="hstack" style=${{ gap: 8, marginTop: 10 }}>
      <${AvatarChip} letter=${a.agentName.charAt(0)} color=${a.agentColor} size=${20} />
      <span class="grow t-xs muted truncate" style=${{ fontSize: 11.5 }}>${a.agentName} · ${a.room} · ${a.age} ago</span>
      <span class="t-xs w8" style=${{ color: a.life.permanent ? "var(--ink-55)" : "var(--tint-amber)", fontSize: 10 }}>${lifeBadge(a.life)}</span>
    </div>
  </div>`;
}
